src/sbsign/sbsign.c

```diff
@@ -800,10 +816,267 @@ static int verb_sign(int argc, char *argv[], void *userdata) {
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read secure boot database signature file '%s': %m", arg_signed_data_signature);
+        }
+
+        _cleanup_free_ uint8_t *dercert = NULL;
+        int dercertsz = i2d_X509(ctx.certificate, &dercert);
+        if (dercertsz < 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to convert X.509 certificate to DER: %s",
+                                        ERR_error_string(ERR_get_error(), NULL));
+
+        uint32_t siglistsz = offsetof(EFI_SIGNATURE_LIST, Signatures) + offsetof(EFI_SIGNATURE_DATA, SignatureData) + dercertsz;
+        /* We use malloc0() to zero-initialize the SignatureOwner field of Signatures[0]. */
+        _cleanup_free_ EFI_SIGNATURE_LIST *siglist = malloc0(siglistsz);
+        if (!siglist)
+                return log_oom();
+
+        *siglist = (EFI_SIGNATURE_LIST) {
+                .SignatureType = EFI_CERT_X509_GUID,
+                .SignatureListSize = siglistsz,
+                .SignatureSize = offsetof(EFI_SIGNATURE_DATA, SignatureData) + dercertsz,
+        };
+
+        memcpy(siglist->Signatures[0].SignatureData, dercert, dercertsz);
+
+        _cleanup_free_ char16_t *db16 = utf8_to_utf16(arg_secure_boot_database, SIZE_MAX);
+        if (!db16)
+                return log_oom();
+
+        EFI_TIME timestamp;
+        if (iovec_is_set(&signed_data)) {
+                /* Don't count the trailing NUL terminator. */
+                size_t db16sz = char16_strsize(db16) - sizeof(char16_t);
+                size_t expectedsz = db16sz + sizeof(EFI_GUID) + sizeof(attrs) + sizeof(timestamp) + siglistsz;
+
+                if (signed_data.iov_len != expectedsz)
+                        return log_error_errno(SYNTHETIC_ERRNO(ERANGE),
+                                               "The secure boot database signed data file size does not match the expected size (%s != %s)",
+                                               FORMAT_BYTES(signed_data.iov_len),
+                                               FORMAT_BYTES(expectedsz));
+
+                /* The signed data includes a timestamp which also has to go in the EFI variable descriptor
+                 * which includes the signature and they have to match, so we extract the timestamp from the
+                 * signed data so we can store it in the EFI variable descriptor later. */
+                size_t tsoffset = db16sz + sizeof(EFI_GUID) + sizeof(attrs);
+                memcpy_safe(&timestamp, (uint8_t*) signed_data.iov_base + tsoffset, sizeof(timestamp));
+        } else {
+                r = efi_timestamp(&timestamp);
+                if (r < 0)
+                        return r;
+        }
+
+        EFI_GUID *guid = STR_IN_SET(arg_secure_boot_database, "PK", "KEK") ? &(EFI_GUID) EFI_GLOBAL_VARIABLE
+                                                                           : &(EFI_GUID) EFI_IMAGE_SECURITY_DATABASE_GUID;
```

Comment:
this makes me feel as if we better should turn this into an enum after all, and then have a table somewhere:

```c
typedef enum SecureBootDatabase {
       SECURE_BOOT_PK,
       SECURE_BOOT_KEK,
       SECURE_BOOT_DB,
       SECURE_BOOT_DBX,
       _SECURE_BOOT_DTABASE_MAX,
       _SECURE_BOOT_DATABASE_INVALID = -EINVAL
} SecureBootDatabase;

static const struct secure_boot_database_table {
        EFI_GUID guid;
        const char *variable_name;
} secure_boot_database_table[] = {
        [SECURE_BOOT_PK] = { EFI_GLOBAL_VARIABLE, "pk" },
        …
}
```

you get the idea...

and then use lowercase enum identifiers everywhere, in sync with out own usual rules for exposed enums. eve if we use uppercase spelling internally for the variables as uefi requires

Reply:
we have a very different table like this already btw, in pcrlock.c, see verb_lock_secureboot_policy().

i think we should maybe unify that, move that to efi-api.[ch] or so. (turn the synthesize_empty field into a simply array in pcrlock.c)

Reply:
there's a 3rd table like this in src/boot/secure-boot.c. but that's harder to unify since it wants utf16, so maybe ignore for now